nex_cv/data_providers.py

```diff
@@ -142,3 +143,162 @@ def __str__(self):
         return super().__str__() + "(%s;%s)" % (self.min_category_size,
                                                 self.other_min_prop)
 
+
+class TimeSeqDatasetProvider(TestingDatasetProvider):
+    """
+    The training data is grown gradually using the creation date for sorting the data
+    "created" is expected to be in the "intent" annotation dict
+    """
+    DATE_KEY = "created"
+    INTENT_KEY = "intent_id"
+
+    def __init__(self, min_examples_ratio=0.3, max_added_examples=None, n_steps=5, date_key=DATE_KEY,
+                 test_on_all=False, test_ratio=0.33):
+        """
+
+        :param min_examples_ratio: (float) ratio of minimum number of training examples
+        :param max_added_examples: (int) maximum number of added examples
+        :param n_steps: (int) number of steps in the learning curve
+        :param date_key: (str) key used for sorting samples
+        :param test_on_all: (bool) if False test set is sampled only from the time covered by training set
+                                  if True test set is sampled from all data
+        :param test_ratio: (float) test data ratio
+        """
+        self.min_examples_ratio = min(min_examples_ratio, 0.8)
+        self.max_added_examples = max_added_examples
+        self.n_steps = max(n_steps, 1)
+
+        self.test_on_all = test_on_all
+        self.test_step = int(1. / test_ratio)
+        self.date_key = date_key
+
+    @staticmethod
+    def resolve_date(a, date_key):
+        """
+
+        :param a: list(dict) with annotations
+        :param date_key: str key used to extract date
+        :return: timestamp extracted from date
+        """
+
+        try:
+            return datetime.timestamp(a.get(date_key))
+        except:
+            return None
+
+    @staticmethod
+    def resolve_intent(a, intent_key):
+        """
+
+        :param a:
+        :param intent_key:
+        :return:
+        """
+        try:
+            return a.get(intent_key)
+        except:
+            return None
+
+    @staticmethod
+    def sort_dataset_by_date(X, y, date_key, return_timestamps=False):
+        """
+
+        :param X: list(str): training sentences
+        :param y: list(dict): with intent_id and created (and updated) dates
+        :param date_key: key used for the date used in sorting the data
+        :param return_timestamps: if the timestamps are returned (e.g. used by the data provider)
+        :return: sorted dataset: X, y[, time_stamps]
+        """
+
+        time_stamps = [TimeSeqDatasetProvider.resolve_date(a, date_key) for a in y]
```

Comment:
What if `resolve_date(..)` returns `None`? Should we just exclude that datapoint, conceptually?

Reply:
The None values are filtered in `sort_dataset_by_date`